q2_metadata/normalization/_norm_check_rules.py

```diff
@@ -0,0 +1,311 @@
+        Parsed rule value for the rule "ontology".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    if isinstance(rule_value, str):
+        if rule_value in ['Gazetteer ontology']:
+            return True
+    return False
+
+
+def check_remap(rule_value):
+    """Check that the user-defined value for the
+    rule "remap" is correctly formatted.
+
+    remap must be a dict
+        Apply the rule of remapping variables values: replace
+        all instances of a given value by another one value.
+                    remap:
+                      str: str
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "remap".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    if isinstance(rule_value, dict):
+        str_values = [
+            x for x, y in rule_value.items() if
+                isinstance(x, (str, int, float)) and
+                isinstance(y, (str, int, float))
+        ]
+        if len(str_values) == len(rule_value.keys()):
+            return True
+    return False
+
+
+def check_validation(rule_value):
+    """Check that the user-defined value for the
+    rule "validation" is correctly formatted.
+
+    validation must be a dict
+        Apply the rule of validating variables values: replace by the blank
+        (or the missing) value every sample entry that for the current variable has
+        in one or multiple other variable(s) a null or missing value for that sample.
+                    validation:
+                      force_to_blank_if:
+                        is null:
+                        - variable
+
+    Parameters
+    ----------
+    rule_value
+        Parsed rule value for the rule "validation".
+
+    Returns
+    -------
+    bool
+        True if successful, False otherwise.
+    """
+    # accepted structure for the "validation" rule
+    possible_validations = {
+        'force_to_blank_if': {'is null'}
+    }
+    # check the first level is a dict
+    if isinstance(rule_value, dict):
+        # get the current keys
+        rule_value_set = set(rule_value)
+        # get the accepted keys from the current keys
+        accepted_set = rule_value_set & set(possible_validations)
+        # if there is not more keys than the accepted ones
+        if not len(rule_value_set ^ accepted_set):
+            # get the possible conditions
+            possible_validations_2 = possible_validations['force_to_blank_if']
+            # for each of the condition/replacements for the accepted keys
+            for rule_2, rule_value_2 in rule_value['force_to_blank_if'].items():
+                # check that the content of replacements are lists
+                if isinstance(rule_value_2, list):
+                    # get the accepted conditions from the current conditions
+                    if rule_2 in possible_validations_2:
+                        return True
+    return False
```

Comment:
The code here is rather hard to follow, also it is in general a bad idea to name variables `rule_2` if there is also  a `rule` variable. Can you think of a way to rewrite this in an simpler form?

Reply:
Rebecca was down to make a recursive function - I can give it a try..

Reply:
If now simplified and more easy to understand:
```
        if isinstance(rule_value, dict):
            if set(rule_value).issubset(allowed):
                for allowed_key, nested_dict in rule_value.items():
                    if isinstance(nested_dict, dict):
                        for nested_rule, variables in nested_dict.items():
                            if nested_rule not in allowed[allowed_key]:
                                return ['not allowed', nested_rule]
                            if not isinstance(variables, list):
                                return ['not a list', variables]
                    else:
                        return ['not a nested dictionary', nested_dict]
            else:
                not_allowed = '", "'.join(set(rule_value).difference(allowed))
                return ['not allowed', not_allowed]
        else:
            return ['not a dictionary', rule_value]
        return []

```